agents/cluster_methods_mixin.py

```
@@ -101,6 +123,40 @@ def _build_cluster_string(
+        empty dict when nothing needs pruning.
+        """
+        ctx = get_current_agent_context_window()
+        prompt_overhead_tokens = prompt_overhead_chars / CHARS_PER_TOKEN
+        available_tokens = (ctx.input_tokens - OUTPUT_HEADROOM_TOKENS - prompt_overhead_tokens) * CONTEXT_MARGIN
+        if available_tokens <= 0:
+            return {}
+        char_budget = int(available_tokens * CHARS_PER_TOKEN)
+        if char_budget <= 0:
+            return {}
+
+        langs_with_clusters = [l for l in programming_langs if cluster_results.get(l)]
```

Ruff (0.15.10)

[error] Ambiguous variable name: `l`

(E741)